Eliminate #any_instance_of in logging specs

imtayadeway opened the pull request:
~~By mocking closer to the boundary (i.e. the `$api_log` itself)~~ By injecting a `StringIO` into the $api_log and making expectations on it, we can decouple these tests from some implementation details that required us to use `#any_instance_of`. The result I think is greater confidence in these tests, greater visibility in what's being logged from the test body, and looser coupling between the test and its implementation, freeing us to refactor some of the logging later if needed. ~~Also, by first stubbing the logger with `allow .... to receive(:info)`, we no longer have to specify every message it receives, giving the tests greater readability.~~ We also no longer required to specify every message that the logger receives (in the right order) - we can instead search through the log device to find only what is relevant to the example at hand. This also removes some constants from the example group which I know will make @kbrock happy :grin: 

@miq-bot add-label api, test, refactoring
@miq-bot assign @abellotti

imtayadeway commented:
Also customary :scissors: :scissors: :scissors:

chessbyte commented:
@abellotti bump

abellotti commented:
wasn't too crazy with the string matching in the specs, vs. the previous hash ways.

imtayadeway commented:
@abellotti would you care to say why? And could you weigh any objections against the benefits I've outlined in https://github.com/ManageIQ/manageiq/pull/15058#issue-227778418 ? I've rebased and updated a little bit since you last looked at it (principally to switch from mocking to verifying against an injected log device), although the expectations remain basically the same. Might allay some of your concerns though as I've tried to make the expectations as order-independent as possible. I think generally the result is a lot better though, because now these tests are testing external behavior, not private implementation details.

kbrock commented:
I like testing at the boundary. And testing in a way that is order independent.
Do agree that hash approach is a little nicer, but this is the logging specs, so testing what goes into the log seems reasonable.

imtayadeway commented:
@abellotti bump

kbrock commented:
The `includes` seems close to the pure hash version now.
Think it is a nice touch
